### ISSUE TITLE
Bullet bug: Assert !fuzzyZero() in line 300, file Bullet/src/LinearMath/btVector3.h

### ISSUE THREAD

**Issue by slapin** [urho3d/Urho3D#1933]
Well, I hit this every time I compile in debug.
This line is annoying and should be deleted. I wonder why it should exit to OS just
because some vector is 0. Otherwise btAssert should be changed not to exit to OS
and just print message (might be dangerous).

Bullet bug: Assert !fuzzyZero() in line 300, file Bullet/src/LinearMath/btVector3.h

```c++
  /**@brief Normalize this vector 
   * x^2 + y^2 + z^2 = 1 */
        SIMD_FORCE_INLINE btVector3& normalize()
        {

                btAssert(!fuzzyZero());
...
       SIMD_FORCE_INLINE bool fuzzyZero() const
        {
                return length2() < SIMD_EPSILON*SIMD_EPSILON;
        }
...
  /**@brief Return the length of the vector squared */
        SIMD_FORCE_INLINE btScalar length2() const
        {
                return dot(*this);
        }
```

That is perfectly stupid stupid stupid...

**Comment by eugeneko** [comment]
> Well, I hit this every time I compile in debug.

You'd probably need to debug your (or Urho's) code. More information needed: callstack, STR, etc.

>  I wonder why it should exit to OS just because some vector is 0.

Broken pre-conidtion is a fault of caller.

> This line is annoying and should be deleted.

Then.. Do you suggest to remove that line?
Is it better to silently explode physical simulation into singularity, IYO?

**Comment by slapin** [comment]
That is bullshit.
The test is error in itself.
If your vector is zero, that is fine, don't do anything to it in normalize().
The crash because of btAssert happens only in debug build. Everything works fine in release.
I assume nobody tested it so nobody have this problem. With this test Urho debug build is broken.
I propose the solution. I don't think you read the code, understand it or care at all. Go do your trolling elsewhere.

**Comment by iSLC** [comment]
Why not open the issue at the official Bullet repository and ask the author why he placed that assert instead of just assuming it's useless based on the fact that it's pissing you off.

**Comment by slapin** [comment]
Because it is useless.
In generic Bullet terms it might be not in case they assume it is not possible to normalize zero vector.
But it perfectly does it (doing nothing) if that is disabled. In debug mode that produce immediate crash.
And no, that thing can't affect anything in simulation. More than that, if it was, it still can't justify crash.
Having zero vector might indicate error, but in case of Urho it is not. And Bullet have very limited information to guess. They guessed it so and have strong belief they are right. But that is wrong for Urho (or anything using Bullet in debug mode).
So the solutions are
1. Go all over Urho code (and everyone else's code) and check that Zero vector is never supplied to normalize(). Which is really stupid as it provdes overhead for nothing.
2. Remove the shitty line and forget about it.
3. Go fight to bullet people, probably win or not, and if win then update Bullet in Urho.
No I don't give a fuck. I have much better things to do.

I disabled the fucking thing in my local copy so I don't care what you do. Consider this as heads up.
As almost nobody compiles Urho in debug mode, that it is probably won't affect anyone anyway, but
I was debugging GUI went haywire and was hit by this too. Please don't comment if you can't say anything intelligent.

**Comment by slapin** [comment]
To simplify - zero vector need not be normalized by normalize(). That doesn't produce any side effects (as it was tested by Urho for a long time). So it is returned as zero vector.

**Comment by cadaver** [comment]
Bullet has a number of asserts, what should we disable in addition to this? Only the Bullet authors know well which is actually safe and won't lead into the simulation exploding later (ie. how is the normalized vector used later? The calling code probably assumes it has non-zero length too.)

When using a library like Bullet, Urho is placed in a kind of undesirable situation. As it shouldn't "ever" crash and any crash / assert would easily be "blamed" on Urho. However, I don't see the assert getting hit in the built-in samples in debug mode and that is enough for me. You're of course free to customize to your liking, but the assert getting hit likely means there's something wrong in your simulation objects.

Finally remember that when you come to the Urho issue tracker, you're a guest, and calling a developer's analysis of the problem "trolling" is not the best course of action and we don't have any obligation to tolerate that.

**Comment by slapin** [comment]
I'm really sorry that the way I speak was possible to interpret as being
rude. I really did not intend to be,
and really never thought to be personal in this case.
In case somebody thought I was being personal and rude, I really sorry
about that and won't post here being
too agitated.

**Comment by eugeneko** [comment]
Finally, do you have some stable STRs?
Such an error may indicate the defect in the Bullet facade provided by Urho, it's a bit uncomfortable for me to just ignore that.
It also may be a defect in some part of the Bullet, so I've asked for callstack.

**Comment by cadaver** [comment]
We can reopen the issue if good evidence materializes, and it's something that is fixable in Urho.

**Comment by Lumak** [comment]
This particular btAssert() should be also enabled for release - it prevents divide by zero.  If you're hitting this assert then you're probably passing a bad vector or corrupt data.

**Comment by slapin** [comment]
The backtrace:
```
#0  normalize (this=<optimized out>) at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/LinearMath/btVector3.h:304
#1  SphereTriangleDetector::collide (this=0x7fffffffb940, sphereCenter=..., point=..., resultNormal=..., depth=@0x7fffffffa638: 0, timeOfImpact=@0x7fffffffa63c: 1, contactBreakingThreshold=0.103923045)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionDispatch/SphereTriangleDetector.cpp:103
#2  0x000000000138792a in SphereTriangleDetector::getClosestPoints (this=0x7fffffffb940, input=..., output=..., debugDraw=0x3115270, swapResults=false)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionDispatch/SphereTriangleDetector.cpp:44
#3  0x000000000138591f in btSphereTriangleCollisionAlgorithm::processCollision (this=0x3645f20, col0Wrap=0x7fffffffc850, col1Wrap=0x7fffffffba80, dispatchInfo=..., resultOut=0x7fffffffc8b0)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionDispatch/btSphereTriangleCollisionAlgorithm.cpp:68
#4  0x000000000135e9e5 in btConvexTriangleCallback::processTriangle (this=0x3645eb0, triangle=0x7fffffffbd80, partId=0, triangleIndex=1045)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionDispatch/btConvexConcaveCollisionAlgorithm.cpp:136
#5  0x0000000000e9dfca in btScaledTriangleCallback::processTriangle (this=0x7fffffffc630, triangle=0x7fffffffc590, partId=0, triangleIndex=1045)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionShapes/btScaledBvhTriangleMeshShape.cpp:50
#6  0x0000000000e6fbd0 in btBvhTriangleMeshShape::MyNodeOverlapCallback::processNode (this=0x7fffffffc570, nodeSubPart=0, nodeTriangleIndex=1045)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionShapes/btBvhTriangleMeshShape.cpp:315
#7  0x0000000000ee6aa5 in btQuantizedBvh::walkStacklessQuantizedTree (this=0x86e5300, nodeCallback=0x7fffffffc570, quantizedQueryAabbMin=0x7fffffffc120, quantizedQueryAabbMax=0x7fffffffc110, startNodeIndex=0, endNodeIndex=10871)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/BroadphaseCollision/btQuantizedBvh.cpp:730
#8  0x0000000000ee2aa3 in btQuantizedBvh::reportAabbOverlappingNodex (this=0x86e5300, nodeCallback=0x7fffffffc570, aabbMin=..., aabbMax=...) at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/BroadphaseCollision/btQuantizedBvh.cpp:346
#9  0x0000000000e6fc53 in btBvhTriangleMeshShape::processAllTriangles (this=0x87a3b10, callback=0x7fffffffc630, aabbMin=..., aabbMax=...) at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionShapes/btBvhTriangleMeshShape.cpp:323
#10 0x0000000000e9c9c6 in btScaledBvhTriangleMeshShape::processAllTriangles (this=0x86e5420, callback=0x3645eb0, aabbMin=..., aabbMax=...)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionShapes/btScaledBvhTriangleMeshShape.cpp:73
#11 0x000000000135fc73 in btConvexConcaveCollisionAlgorithm::processCollision (this=0x3645e90, body0Wrap=0x7fffffffc850, body1Wrap=0x7fffffffc880, dispatchInfo=..., resultOut=0x7fffffffc8b0)
    at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionDispatch/btConvexConcaveCollisionAlgorithm.cpp:208
#12 0x0000000000ed36f8 in btSingleContactCallback::process (this=0x7fffffffce80, proxy=0x88d90b0) at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionDispatch/btCollisionWorld.cpp:1157
#13 0x0000000000efbc2e in BroadphaseAabbTester::Process (this=0x7fffffffce30, leaf=0x88dc220) at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/BroadphaseCollision/btDbvtBroadphase.cpp:264
#14 0x0000000000efa6ce in btDbvt::collideTV (this=0x30c53f8, root=0x86024a0, vol=..., policy=...) at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/BroadphaseCollision/btDbvt.h:946
#15 0x0000000000ef5d8d in btDbvtBroadphase::aabbTest (this=0x30c53f0, aabbMin=..., aabbMax=..., aabbCallback=...) at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/BroadphaseCollision/btDbvtBroadphase.cpp:274
#16 0x0000000000ec7c79 in btCollisionWorld::contactTest (this=0x314fce0, colObj=0x355e300, resultCallback=...) at /home/slapin/Urho3D/Source/ThirdParty/Bullet/src/BulletCollision/CollisionDispatch/btCollisionWorld.cpp:1176
#17 0x0000000000c20241 in Urho3D::PhysicsWorld::GetRigidBodies (this=0x31151e0, result=..., sphere=..., collisionMask=12) at /home/slapin/Urho3D/Source/Urho3D/Physics/PhysicsWorld.cpp:623
#18 0x00000000009abeb5 in Urho3D::PhysicsWorldGetRigidBodiesSphere (sphere=..., collisionMask=12, ptr=0x31151e0) at /home/slapin/Urho3D/Source/Urho3D/AngelScript/PhysicsAPI.cpp:273
#19 0x0000000000f447d3 in endstack () at /home/slapin/Urho3D/Source/ThirdParty/AngelScript/source/as_callfunc_x64_gcc.cpp:162
#20 0x0000000000f451ca in CallSystemFunctionNative (context=0x30db090, descr=0x2b04b50, obj=0x31151e0, args=0x2f606dc, retPointer=0x0, retQW2=@0x7fffffffd5a0: 0, secondObject=0x0)
    at /home/slapin/Urho3D/Source/ThirdParty/AngelScript/source/as_callfunc_x64_gcc.cpp:468
#21 0x0000000000904791 in CallSystemFunction (id=12261, context=0x30db090) at /home/slapin/Urho3D/Source/ThirdParty/AngelScript/source/as_callfunc.cpp:712
#22 0x00000000008eb68c in asCContext::ExecuteNext (this=0x30db090) at /home/slapin/Urho3D/Source/ThirdParty/AngelScript/source/as_context.cpp:2515
#23 0x00000000008e8b7d in asCContext::Execute (this=0x30db090) at /home/slapin/Urho3D/Source/ThirdParty/AngelScript/source/as_context.cpp:1297
#24 0x00000000009c8d37 in Urho3D::ScriptFile::Execute (this=0x2f4dff0, object=0x8c879a0, method=0x30e65e0, parameters=..., unprepare=true) at /home/slapin/Urho3D/Source/Urho3D/AngelScript/ScriptFile.cpp:366
#25 0x0000000000993afa in Urho3D::ScriptInstance::HandleSceneUpdate (this=0x8c2d9b0, eventType=..., eventData=...) at /home/slapin/Urho3D/Source/Urho3D/AngelScript/ScriptInstance.cpp:839
#26 0x00000000009981f4 in Urho3D::EventHandlerImpl<Urho3D::ScriptInstance>::Invoke (this=0x8c87650, eventData=...) at /home/slapin/Urho3D/Source/Urho3D/AngelScript/../AngelScript/../Core/Object.h:307
#27 0x0000000000b890d8 in Urho3D::Object::OnEvent (this=0x8c2d9b0, sender=0x3009a40, eventType=..., eventData=...) at /home/slapin/Urho3D/Source/Urho3D/Core/Object.cpp:113
#28 0x0000000000b89b16 in Urho3D::Object::SendEvent (this=0x3009a40, eventType=..., eventData=...) at /home/slapin/Urho3D/Source/Urho3D/Core/Object.cpp:325
#29 0x0000000001091c95 in Urho3D::Scene::Update (this=0x3009a40, timeStep=0.0917575061) at /home/slapin/Urho3D/Source/Urho3D/Scene/Scene.cpp:781
#30 0x0000000001093731 in Urho3D::Scene::HandleUpdate (this=0x3009a40, eventType=..., eventData=...) at /home/slapin/Urho3D/Source/Urho3D/Scene/Scene.cpp:1171
#31 0x0000000001099676 in Urho3D::EventHandlerImpl<Urho3D::Scene>::Invoke (this=0x30b2f00, eventData=...) at /home/slapin/Urho3D/Source/Urho3D/Scene/../Core/../Core/Object.h:307
#32 0x0000000000b89123 in Urho3D::Object::OnEvent (this=0x3009a40, sender=0x1ce3dd0, eventType=..., eventData=...) at /home/slapin/Urho3D/Source/Urho3D/Core/Object.cpp:121
#33 0x0000000000b89cb8 in Urho3D::Object::SendEvent (this=0x1ce3dd0, eventType=..., eventData=...) at /home/slapin/Urho3D/Source/Urho3D/Core/Object.cpp:355
#34 0x0000000000ba8622 in Urho3D::Engine::Update (this=0x1ce3dd0) at /home/slapin/Urho3D/Source/Urho3D/Engine/Engine.cpp:695
#35 0x0000000000ba7f62 in Urho3D::Engine::RunFrame (this=0x1ce3dd0) at /home/slapin/Urho3D/Source/Urho3D/Engine/Engine.cpp:519
#36 0x0000000000ba4b06 in Urho3D::Application::Run (this=0x1ce3c40) at /home/slapin/Urho3D/Source/Urho3D/Engine/Application.cpp:86
#37 0x00000000008e2a5b in RunApplication () at /home/slapin/dungeon/rework/Urho3DPlayer.cpp:47
#38 0x00000000008e2b02 in main (argc=1, argv=0x7fffffffe248) at /home/slapin/dungeon/rework/Urho3DPlayer.cpp:47
```
Feel free to reopen, if you want.

**Comment by cadaver** [comment]
Is there an easy way to repro with Urho's builtin samples?

**Comment by slapin** [comment]
Not that I know of.

**Comment by cadaver** [comment]
I'll still test with GetRigidBodies (sphere overload), if it happens in Urho physics example scenes then it certainly warrants reopening.

**Comment by eugeneko** [comment]
It looks like your triangle mesh have degenerate triangles.
What trimesh do you use for collision?

**Comment by slapin** [comment]
I export from Blender and also generate some geometry.
Thanks for idea. Will try to find which geometry causes it.
But I wonder if a few checks in Urho would hurt too much performance in
debug mode (as it crashes only in debug mode).

**Comment by eugeneko** [comment]
This issue should probably be solved by Urho-side check of bad trimeshes, probably warning+removing degenerates or just ignoring bad mesh.

**Comment by cadaver** [comment]
Worth reopening.

**Comment by Lumak** [comment]
Whatever you do, I request that Bullet's assert in debug not be removed. It's critical to have that for development sake. Thx!

**Comment by eugeneko** [comment]
@Lumak of course it won't. This assert is fine and caught at least one bug.

**Comment by Lumak** [comment]
I'm relieved to hear that. Thx.

**Comment by slapin** [comment]
Closed via 86ba5e5bdd1ebad423935b3289e21c0ec49fe807
